File: examples/csharp/Patterns/AbstractFactory.cs
namespace DesignPatterns.AbstractFactory
{
    public interface IButton { string Render(); }
    public interface ICheckbox { string Render(); }

    public interface IGUIFactory
    {
        IButton CreateButton();
        ICheckbox CreateCheckbox();
    }

    public class WindowsButton : IButton { public string Render() => "Rendering Windows button"; }
    public class WindowsCheckbox : ICheckbox { public string Render() => "Rendering Windows checkbox"; }

    public class LinuxButton : IButton { public string Render() => "Rendering Linux button"; }
    public class LinuxCheckbox : ICheckbox { public string Render() => "Rendering Linux checkbox"; }

    public class WindowsFactory : IGUIFactory
    {
        public IButton CreateButton() => new WindowsButton();
        public ICheckbox CreateCheckbox() => new WindowsCheckbox();
    }

    public class LinuxFactory : IGUIFactory
    {
        public IButton CreateButton() => new LinuxButton();
        public ICheckbox CreateCheckbox() => new LinuxCheckbox();
    }

    public class Application
    {
        private readonly IButton _button;
        private readonly ICheckbox _checkbox;

        public Application(IGUIFactory factory)
        {
            _button = factory.CreateButton();
            _checkbox = factory.CreateCheckbox();
        }

        public void Render()
        {
            Console.WriteLine(_button.Render());
            Console.WriteLine(_checkbox.Render());
        }
    }

    public static class Demo
    {
        public static void Run()
        {
            Console.WriteLine("--- Windows GUI ---");
            new Application(new WindowsFactory()).Render();

            Console.WriteLine("--- Linux GUI ---");
            new Application(new LinuxFactory()).Render();
        }
    }
}

File: examples/cpp/abstract-factory.cpp
#include <iostream>
#include <memory>
#include <string>

// Abstract products
class Button {
public:
    virtual ~Button() = default;
    virtual std::string render() const = 0;
};

class Checkbox {
public:
    virtual ~Checkbox() = default;
    virtual std::string render() const = 0;
};

// Windows variants
class WindowsButton : public Button {
public:
    std::string render() const override { return "Rendering Windows button"; }
};

class WindowsCheckbox : public Checkbox {
public:
    std::string render() const override { return "Rendering Windows checkbox"; }
};

// Linux variants
class LinuxButton : public Button {
public:
    std::string render() const override { return "Rendering Linux button"; }
};

class LinuxCheckbox : public Checkbox {
public:
    std::string render() const override { return "Rendering Linux checkbox"; }
};

// Abstract factory
class GUIFactory {
public:
    virtual ~GUIFactory() = default;
    virtual std::unique_ptr<Button> createButton() const = 0;
    virtual std::unique_ptr<Checkbox> createCheckbox() const = 0;
};

class WindowsFactory : public GUIFactory {
public:
    std::unique_ptr<Button> createButton() const override {
        return std::make_unique<WindowsButton>();
    }
    std::unique_ptr<Checkbox> createCheckbox() const override {
        return std::make_unique<WindowsCheckbox>();
    }
};

class LinuxFactory : public GUIFactory {
public:
    std::unique_ptr<Button> createButton() const override {
        return std::make_unique<LinuxButton>();
    }
    std::unique_ptr<Checkbox> createCheckbox() const override {
        return std::make_unique<LinuxCheckbox>();
    }
};

void buildUI(const GUIFactory& factory) {
    auto button = factory.createButton();
    auto checkbox = factory.createCheckbox();
    std::cout << button->render() << std::endl;
    std::cout << checkbox->render() << std::endl;
}

int main() {
    std::cout << "=== Abstract Factory: GUI Widgets ===" << std::endl;

    std::cout << "\nWindows GUI:" << std::endl;
    WindowsFactory winFactory;
    buildUI(winFactory);

    std::cout << "\nLinux GUI:" << std::endl;
    LinuxFactory linFactory;
    buildUI(linFactory);

    return 0;
}
